Student: feedback submission: distribution points question: explain if equal points are not allowed #8817

mg14777 opened the pull request:
Fixes #8817 
Added Further UI Enhancements

**Outline of Solution**
-  All the constraints to fulfilled while distributing points are now displayed as information points
- The user is informed in real time if he has satisfied all the constraints, after which he can submit feedback
- The feature has been implemented for both, distribute points `among recipients` and `among options` type of questions

**Explanation**
If the question puts a constraint on total points and uneven distribution of points across all users the user is constantly informed which constraints are satisfied so far

- Complete uneven distribution
![distribute-unequal-final-wrong-2](https://user-images.githubusercontent.com/11634699/40188220-5e9a24ee-59f1-11e8-9738-29012b37658c.PNG)
![distribute-unequal-final-wrong-0](https://user-images.githubusercontent.com/11634699/40188345-a3170f1a-59f1-11e8-9cc4-968ca8fa3bdf.PNG)

- Some uneven distribution
![distribute-options-final-wrong-2](https://user-images.githubusercontent.com/11634699/40188679-6419d580-59f2-11e8-9541-33a8895f1461.PNG)
![distribute-options-final-wrong-0](https://user-images.githubusercontent.com/11634699/40188690-6bdc3f06-59f2-11e8-9109-f0884dcd9cd6.PNG)

**Clarifications**
Not sure regarding the difference between variables `forceUnevenDistribution` and `forceAllUnevenDistribution`. Also for `among recipients ` type questions, both `allowEvenDistribution` and `forceUnevenDistribution` were true simultaneously.

<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->

mg14777 commented:
@wenmogu @joanneong  Ready for review
@damithc I have changed the status messages to constantly inform user for both instructions

sukanta-27 commented:
@mg14777 Just an advise for your future PRs. Your PR title should be <em>exact</em> same as your issue title. Until you are fixing an issue that needs more than one PR to fix.

wenmogu commented:
I will take a closer look at the code later but just wanna express my opinion about the UI. It is simple and clear and I like the color combination, but as something displaying data I feel it's a bit too wordy. The user might want the information in one glance. My suggestion is something like
``` java
expected total:
actual total: (the number turns red if it exceeds)

expected mode: different values for all inputs
```
You shld also shift the expected mode to the front, so that the user can see if before he does the survey. Sometimes the user might have to scroll down to see the submit button.

damithc commented:
> I think its good that theres a reminder about the mode together with other reminders. but you shld also have one mode reminder in the front, so that the user can see if before he does the survey. Sometimes the user might have to scroll down to see the submit button.

This is a good point. May be the blue instructions can be shown above the input boxes?

As for wordiness, yes it is a bit verbose but I'd rather be safe. Some of the TEAMMATES users may need detailed explanation to understand what's going on. Also, we can't ignore the fact that a significant portion of the population are color blind, especially between blue and red.

wenmogu commented:
> As for wordiness, yes it is a bit verbose but I'd rather be safe. Some of the TEAMMATES users may need detailed explanation to understand what's going on. 

What about hiding those detailed explanations inside a box that only appears when the mouse hovers over the info icon? This also aligns with the current UI design for many other pages.

> Also, we can't ignore the fact that a significant portion of the population are color blind, especially between blue and red.

what about something like this:
```java
expected total:
actual total: (changes to word reminder only when it exceeds expected total)

mode: different points for all inputs
```

damithc commented:
It is true that the current phrases are a bit too verbose; we can try to make the phrases more concise once the PR is in. For wording tweaks, it's easier to comment on the code than screenshots.
Furthermore, we can probably omit the green text altogether as only errors need to be communicated to the user.

> What about hiding those detailed explanations inside a box that only appears when the mouse hovers over the info icon? This also aligns with the current UI design for many other pages.

This is a case where direct communication is needed. 'hover' is good for optional additional info. For non-tech users, I'd rather explain the situation as a sentence or a phrase than give 'expected' and 'actual' 
 and let the user figure out the implication.

mg14777 commented:
> I think its good that theres a reminder about the mode together with other reminders. but you shld also have one mode reminder in the front, so that the user can see if before he does the survey. Sometimes the user might have to scroll down to see the submit button.

> This is a good point. May be the blue instructions can be shown above the input boxes?
I also agree with this. Maybe I can go ahead and make this change since we all agree on this.

> Furthermore, we can probably omit the green text altogether as only errors need to be communicated to the user.
Or maybe just display a single message saying `All checks passed` ?

I was wondering if I can go ahead and modify the browser tests or do we still plan on making further UI changes?

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -518,32 +516,40 @@ function updateConstSumMessageQn(qnNum) {
-                    || (forceAllUnevenDistribution && allUnique)
-                    || (forceSomeUnevenDistribution && !allSame)) {
-                message = 'All points distributed!';
+        if (!allNotNumbers) {
```
just saying if there are a lot of double negations for this variable, then you can consider changing the variable name to prevent such double negation.

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -518,32 +516,40 @@ function updateConstSumMessageQn(qnNum) {
-            if (allowEvenDistribution || !forceUnevenDistribution
-                    || (forceUnevenDistribution && !forceSomeUnevenDistribution && allUnique)
-                    || (forceAllUnevenDistribution && allUnique)
-                    || (forceSomeUnevenDistribution && !allSame)) {
```
`allowEvenDistribution` means some or all options can have the same points
`forceUnevenDistribution` means there have to be at least 2 options with different points
`forceAllUnevenDistribution` means everything has to be different
`forceSomeUnevenDistribution` is the intersection of `allowEvenDistribution`  and `forceUnevenDistribution`. 
As u can see, `forceSomeUnevenDistribution` can be true. So  `allowEvenDistribution`  and `forceUnevenDistribution` can be true at the same time.

I think you shld learn from this code fragment as you can tell each line i.e `allowEvenDistribution || !forceUnevenDistribution`, 
`(forceUnevenDistribution && !forceSomeUnevenDistribution && allUnique)`, `(forceAllUnevenDistribution && allUnique)` and `(forceSomeUnevenDistribution && !allSame)` each of them is a case of user input. Writing code is way is neater and easier for other contributors to read.

mg14777 replied:
Thanks for such a detailed explanation. Really helpful. I was performing a few experiments to get my head around this. I had some questions if that's alright.

To recap my understanding (please correct me if I am wrong):
- `forceSomeUnevenDistribution`: Implies there has to be a mixture, all can't be the same and all can't be different as well
- In `allowEvenDistribution`: at least two options have to be the same ? i.e. it is exclusive with`forceAllUnevenDistribution`
- When `forceSomeUnevenDistribution` is true both `allowEvenDistribution` and `forceUnevenDistribution` have to be true since it is the intersection

So first of all lets consider the `among recipients` question as given in one of the screenshots (question no. 6). When I created this question I selected the option `Every option to receive a different number of points`. Based on my understanding that would imply `forceAllUnevenDistribution`. I tried to print all these variables to the console for question 6 shown below:

`Question:  6`
`Allow even distribution:  true`
`Force uneven distribution:  true`
`Force some uneven distribution:  false`
`Force all uneven distribution:  false`

`forceAllUnevenDistribution` is not true against what I thought and same goes for `forceSomeUnevenDistribution`. Am i missing something here ?

Finally from the code fragment above aren't the conditions `(forceUnevenDistribution && !forceSomeUnevenDistribution && allUnique)` and `(forceAllUnevenDistribution && allUnique)` the same thing ?

mg14777 replied:
Similarly for question 7 I selected `At least some options to receive a different number of points`. I thought that basically means all can't be the same which implies `forceAllUnevenDistribution`, `forceSomeUnevenDistribution` and `forceUnevenDistribution` should be true. Only `allowEvenDistribution` should be false.

wenmogu replied:
hello @mg14777 , so sorry for the super late reply. I was quite caught up with my course. 

For the concepts, sorry i did not give a clearer explanation which led to your confusion. Thank you for clarifying with me. I think I gave some very confusing answer.

Now let me answer this again. I think somehow you have already figured some out. 

>forceSomeUnevenDistribution: Implies there has to be a mixture, all can't be the same and all can't be different as well
In allowEvenDistribution: at least two options have to be the same ? i.e. it is exclusive withforceAllUnevenDistribution
When forceSomeUnevenDistribution is true both allowEvenDistribution and forceUnevenDistribution have to be true since it is the intersection

`forceSomeUnevenDistribution` in the `Const.js` file, the explanation is 'At least some options'. This means all can't be the same, but all can be different.
`allowEvenDistribution` corresponds to not ticking the checkbox of "every option/at least some options to receive a different number of points", which means, the student can fill up in whatever way they want, without any restriction besides hving to reach the total points,

Basically, `forceSomeUnevenDistribution` ,  `allowEvenDistribution`  and `forceAllUnevenDistribution` are the three scenarios. You can say they are mutually exclusive in the sense that only one out of the three can be true, as you can tell that these boolean values are gotten by seeing if they match the string values.

```java
const allowEvenDistribution = constSumDistributePointsOptionSelected
            === ParamsNames.FEEDBACK_QUESTION_CONSTSUMNOUNEVENDISTRIBUTION;
    const forceAllUnevenDistribution = constSumDistributePointsOptionSelected
            === ParamsNames.FEEDBACK_QUESTION_CONSTSUMALLUNEVENDISTRIBUTION;
    const forceSomeUnevenDistribution = constSumDistributePointsOptionSelected
            === ParamsNames.FEEDBACK_QUESTION_CONSTSUMSOMEUNEVENDISTRIBUTION;
```

If you open the distribution questions from the student side, or `preview` as student, when you inspect the distribution questions, you will see a lot of hidden inputs inside each question. 
![image](https://user-images.githubusercontent.com/28526038/40306253-8be14b0c-5d30-11e8-8e99-b27a5853977a.png)

These hidden inputs are the indicators/flags for this question regarding its setting. You can find the use of these flags in method `updateConstSumMessageQn(qnNum)` in `feedbackSubmissionsEdit.js`.

In summary (sorry i m being so long winded), 

When `forceSomeUnevenDistribution` is true,  `allowEvenDistribution` is false, and `forceAllUnevenDistribution` is false,  and `forceUnevenDistribution` is false. 

When `allowEvenDistribution` is true, `forceSomeUnevenDistribution` and `forceAllUnevenDistribution` is false, and `forceUnevenDistribution` is false.

When `forceAllUnevenDistribution` is true, the other two cases are false and `forceUnevenDistribution` is false.

wenmogu replied:
as for this
>Finally from the code fragment above aren't the conditions (forceUnevenDistribution && !forceSomeUnevenDistribution && allUnique) and (forceAllUnevenDistribution && allUnique) the same thing ?

yeah they are the same thing. Firstly I suggest you not to use `forceUnevenDistribution`, as it is equivalent to `forceAllUnevenDistribution`. I think it's not cleared in the last PR regarding this. Our bad. But dont clear it in this PR, as this is not the concern of your issue. 

My concern is just out of readability. When there are too many booleans chained especially with negation, it's a bit more difficult to understand and get the logic right for those who read the code. We should aim to keep our code as simple as possible:))))

mg14777 replied:
@wenmogu Thanks a lot for clarifying my doubts. It's alright. I can imagine you must be busy. Apologies from my side as well for not being active for the past two days. I was partly awaiting clarification and partly preparing for the interview. But it's done now so it should be fine.

So in a nutshell, based on our discussion so far, I will try to keep the boolean expressions as simple as possible. I think there is still some confusion regarding mutual exclusiveness of the variables.
> Firstly I suggest you not to use forceUnevenDistribution, as it is equivalent to forceAllUnevenDistribution.

My experiment as mentioned above has shown otherwise. `forceUnevenDistribution` is true while `forceAllUnevenDistribution` is false when `Every option to receive a different number of points` is checked. Anyway, I will look into this to see why this is happening. 

I was just thinking instead of all this why don't we just have two variables (`forceAllUnevenDistribution` and `forceSomeUnevenDistribution`) corresponding to the two different constraints that could be checked. I won't modify anything right now but this is just a suggestion.

wenmogu replied:
>I was just thinking instead of all this why don't we just have two variables (forceAllUnevenDistribution and forceSomeUnevenDistribution) corresponding to the two different constraints that could be checked. I won't modify anything right now but this is just a suggestion.

I think you can do it this way, since if theres no restriction on input then theres no need to set any reminder for it.

damithc commented:
> I also agree with this. Maybe I can go ahead and make this change since we all agree on this.

Go ahead.

May be you can stage the latest changes in your staging server so that we can play around with it?

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -518,32 +516,40 @@ function updateConstSumMessageQn(qnNum) {
     }
 
-    function checkAndDisplayMessage(messageElement) {
+    function checkAndDisplayMessage(instructionElement, messageElement) {
```
Absolutely

mg14777 commented:
@damithc @wenmogu Following changes have been made:

- Instructions now displayed above as notes
- Names/ labels aligned in straight line with the input boxes to reduce form filling time

Link to staging server: https://teammates-mudit.appspot.com/

![distribute-recipients-1-1](https://user-images.githubusercontent.com/11634699/40428530-656babfc-5e98-11e8-8ac3-1fe312aed6f1.PNG)

**Problems faced**
Spent a lot of time figuring out the best way to put instructions above the question. Couldn't do it without putting the markup in .tag file instead of html template since everything above the form is in the .tag file. The approach I am using is to add empty markup in the .tag file and then manipulate it using JS (both instructions and error messages).

**Bugs**
As specified by @wenmogu I should ideally be using `forceAllUnevenDistribution` instead of `forceUnevenDistribution` but there's something wrong because for Question 6 I checked `Every option to receive a different number of points` but `forceUnevenDistribution` variable is still false because `constSumDistributePointsOptionSelected` is `None` while it should ideally be `All options`
I am looking into this atm.

Meanwhile I am making some more changes to the UI. Please let me know any feedback regarding the current version.

mg14777 commented:
Bug mentioned above has been fixed. I made some changes to indentation of `distribute among options` style questions to handle multiple parameters for multiple students. It looks like this now:

![image](https://user-images.githubusercontent.com/11634699/40446802-5fc72b9c-5ec8-11e8-8fd2-f11b33dccbdd.png)

Ready for review

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -560,35 +594,45 @@ function updateConstSumMessageQn(qnNum) {
-            messageElement.addClass('text-color-red');
-            messageElement.removeClass('text-color-green');
+            if (forceSomeUnevenDistribution) {
+                if (allSame && numOptions !== 1) {
```
forgot to say that i like how u separated this part of logic into nested if-else blocks. It looks simpler now.

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/questionWithResponses.tag:
```diff
@@ -56,7 +58,11 @@
+          <p class="text-color-blue align-left text-bold">Note:</p>
+          <p class="text-color-blue padding-left-35px" id="constSumInstruction-${questionWithResponses.question.questionNumber}"></p>
+        <hr class="margin-top-0 border-top-dark-gray">
+      </c:if>
```
Hmm ok I'll move everything to JS side

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -479,24 +479,61 @@ function prepareConstSumQuestions() {
         }
     }
 }
+function updateConstSumInstructionsQn(qnNum) {
```
Sure. I was thinking `updateConstSumQnInstructions` and `updateConstSumQnMessage` might be appropriate ?

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -479,24 +479,61 @@ function prepareConstSumQuestions() {
     const numRecipients = parseInt($(`[name="questionresponsetotal-${qnNum}"]`).val(), 10);
     const distributeToRecipients = $(`#constSumToRecipients-${qnNum}`).val() === 'true';
     const pointsPerOption = $(`#constSumPointsPerOption-${qnNum}`).val() === 'true';
-    const forceUnevenDistribution = $(`#constSumUnevenDistribution-${qnNum}`).val() === 'true';
```
Ok i'll look into this

mg14777 replied:
So I went through the PR which was helpful. Based on my understanding, `forceUnevenDistribution` was the variable used to mention `All options are different` before `Some options` feature came along. I think the reason the reminders aren't working in the old version is because I have completely eliminated that variable in my version. So I'll include it as well in the if conditions to ensure it is backward compatible. Hopefully that should resolve it

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -479,24 +479,61 @@ function prepareConstSumQuestions() {
+            } else {
+                $(`[id^="constSumInfo-${qnNum}-"]`).show();
+                $(`[id^="constSumSubmissionFormOptionFragment-${qnNum}"]`).addClass('padding-left-55px');
+                $(`[id^="constSumSubmissionFormOptionFragment-${qnNum}"]`).addClass('margin-top-15px');
```
@wenmogu  The reason behind further indentation in options style questions is because those input boxes are parameters that are evaluated for each student separately. Those don't belong to `Evaluee/Recipient`. In the above view 11, 22, etc. are parameters that are evaluated for each recipient. You should have another heading above that listing the recipient name ( A, B, etc.). Not 
sure why it isn't there. Then the alignment will look appropriate like this:
![image](https://user-images.githubusercontent.com/11634699/40487483-331851a4-5f5c-11e8-87d0-4a7ae39d70c0.png)

If I put Grades, Sports, etc. under the same column as Student names it looks confusing

mg14777 commented:
@wenmogu @damithc     All checks passed. Ready for review.


**Problems faced/ Questions**
While trying to fix `AdminHomePageUiTest` I was facing this problem while testing locally. The instructor join link is generated using:

`// use AppUrl from Config because the join link takes its base URL from build.properties`
`String expectedjoinUrl = Config.getAppUrl(Const.ActionURIs.INSTRUCTOR_COURSE_JOIN)`
                                        `.withRegistrationKey(encryptedKey)`
                                        `.withInstructorInstitution(institute)`
                                        `.toAbsoluteString();`

I had two questions here. First, the comment above mentions to take the AppUrl from build.properties however in the Config file the AppUrl is obtained via GAE's environment object while all other parameters are obtained from `build.properties`. Why is this ?

![image](https://user-images.githubusercontent.com/11634699/40611694-58752bfa-626f-11e8-817a-64cc42c38609.png)

Second, the base URL (`APP_URL`) returned is null instead of localhost:8080 since it is retrieved from the `com.google.appengine.runtime.default_version_hostname` attribute of current environment which is not present. Not sure why is that ?

![image](https://user-images.githubusercontent.com/11634699/40611818-c42c4dc4-626f-11e8-90d0-8bc0bcbd2211.png)

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -474,19 +474,74 @@ function prepareConstSumQuestions() {
         }
     }
 }
+function updateConstSumQnInstructions(qnNum) {
```
uhmmm there are a lot of repetition between this function and the `updateConstSumQnMessage`. Is there a way to reduce such repetition?

mg14777 replied:
Yes there's a bit of repetition here. Initially the code of this function was part of the message function so there was no repetition. I moved it into another function since the instructions need to be updated only once at the time page is loaded while the messages are updated at every user input. 

Maybe we can use a JS module ? 
const constSumQuestions = (function() {
       // constant declarations
       const ....
      // methods
      var instructions = function () { ... };
      var messages = functions () { ... };

      return {
           instructions: instructions,
           messages: messages
      }
}());

wenmogu replied:
uhmmm just see what u can do...the readability is the priority so if after the extraction its not very readable, we can give up on this idea

wenmogu commented on src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSubmitPageUiTest.java:
```diff
@@ -464,21 +464,50 @@ private void testConstSumSubmitAction() {
+        qnNumber = 22;
+        assertEquals("Total points distributed should add up to 100.\n"
+                    + "At least one option should be allocated different number of points.",
+                    submitPage.getConstSumInstruction(qnNumber));
```
do the above questions include const sum option questions?

wenmogu replied:
then how come the tests can pass for the const sum option questions if these instructions dont appear for those questions without recipient names?

mg14777 replied:
the tests are passing because almost all the questions have the recipient names (for options). I didn't consider any new questions. I augmented the tests for the existing const sum questions that were being tested as part of method `testConstSumSubmitAction`. Question 17 is the only question without the recipient name and I missed it so I'll add a test for that

mg14777 replied:
While we are on that subject I was trying to understand why do the question numbers differ between `InstructorFeedbackSubmitPageUiTest.json` and the actual page. I thought the test data is extracted from the json file so it should be exactly the same as what is displayed ?

nidhi98gupta commented:
@mg14777 Hi, Few of my tests were also failing because of `Malformed Url Exception`. What did you do to resolve this error?


> Second, the base URL (APP_URL) returned is null instead of localhost:8080 since it is retrieved from the com.google.appengine.runtime.default_version_hostname attribute of current environment which is not present.

Same was happening in my case too.

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/questionWithResponses.tag:
```diff
@@ -1,9 +1,11 @@
 <%@ tag trimDirectiveWhitespaces="true" %>
 <%@ tag description="feedbackSubmissionEdit.jsp - Display question with responses" pageEncoding="UTF-8" %>
 <%@ tag import="teammates.common.util.Const"%>
+<%@ tag import="teammates.common.datatransfer.questions.FeedbackQuestionType"%>
```
that's redundant. thanks for pointing that out!

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/questionWithResponses.tag:
```diff
@@ -1,9 +1,11 @@
 <%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
 <%@ taglib tagdir="/WEB-INF/tags/shared/feedbackSubmissionEdit" prefix="feedbackSubmissionEdit" %>
 
+<%@ attribute name="feedbackConstSumQuestionDetails" type="teammates.common.datatransfer.questions.FeedbackConstantSumQuestionDetails"%>
```
yes that's right

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/questionWithResponses.tag:
```diff
@@ -56,9 +58,8 @@
-
       <c:if test="${not isRecipientNameHidden}">
-        <div class="col-sm-12 form-inline mobile-align-left">
+        <div class="evalueeLabel-${questionWithResponses.question.qnIndx} col-sm-12 form-inline mobile-align-left">
```
I named it evalueeLabel since it is a static label that says Evaluee/Recipient. Maybe you have a better name in mind ? 

Yes I was thinking about this. The instructions won't appear if recipient's name is hidden. Currently the only thing I can of is having an empty div element particularly dedicated for instructions above the evalueeLabel div to completely eliminate dependence on this one. I can then manipulate it using JS ?

mg14777 replied:
Ok so I think it might be beneficial to add an empty div for instructions due to 2 reasons. First, we will no longer depend on the recipient for manipulating it. Second the requirement of instructions is not just for const sum questions. This feature can be extended to include other question types as well eg. rubric. So I think it will be a better choice in the longer run. Let me know what you think. I don't want to make changes before your approval

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag:
```diff
@@ -23,9 +23,9 @@
-  <div ${isNumResponsesMax ? 'class="col-sm-3 form-inline mobile-align-left"' : 'class="col-sm-5 form-inline mobile-align-left"'}
-      ${isRecipientNameHidden ?  'style="display:none"' : 'style="text-align:right"'}>
+<div class="evalueeForm-${questionWithResponses.question.qnIndx} form-group margin-0" style="padding-left: 75px">
+  <div ${isNumResponsesMax ? 'class="col-sm-12 form-inline mobile-align-left"' : 'class="col-sm-5 form-inline mobile-align-left"'}
```
you mean the name evalueeForm ? Well these elements are part of the form for different evaluees that's why ? I think col-sm-12 is useful here since it adds padding to ensure the text box and the recipient name are in line. I tested it. After removing col-sm-12 for second recipient it looks like this:

![image](https://user-images.githubusercontent.com/11634699/40686754-5149470e-6390-11e8-8a6f-d6490514eba1.png)

mg14777 commented:
@nidhi98gupta Hi Nidhi, thanks for reaching out. I used a simple hack and changed the line:
`APP_URL = readAppUrl()` to `APP_URL = properties.getProperty("app.url")` so that it picks it from `build.properties`. Make sure to mention `localhost:8080` for app.url in `build.properties`. The test ran successfully after that but I reverted the line back to its original version.

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/questionWithResponses.tag:
```diff
@@ -56,9 +56,8 @@
-
       <c:if test="${not isRecipientNameHidden}">
-        <div class="col-sm-12 form-inline mobile-align-left">
+        <div class="evalueeLabel-${questionWithResponses.question.qnIndx} form-inline mobile-align-left">
```
i do add the instructions outside this block atm. It looks something like this:
![image](https://user-images.githubusercontent.com/11634699/40718749-1bfcffea-6409-11e8-8a00-4165fadfa6cc.png)

As you can see the instructions are added outside the `evalueeLabel` div. The problem is that I need that div as a hook  to add my elements before it (using `before` funciton). If recipient names are hidden the div is not there and hence the instructions are not displayed. 

That why I think I will simply add an empty div before that if condition of the form:
`<div class="constraints-${questionWithResponses.question.qnIndx}"></div>`
and use it as the new hook. Does that make sense ?

wenmogu replied:
uhmmm hv u considered using a div just to contain the instructions? using something like [prepend](http://api.jquery.com/prepend/) so that this instruction div will be independent of this if condition?

mg14777 replied:
yes that's what i have done in my latest commit [c4f442b](8836/commits/c4f442b74c3e604637d9b66eeaa4130fca32672e)

mg14777 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/questionWithResponses.tag:
```diff
@@ -56,9 +56,9 @@
         </ul>
       </c:if>
-
+      <div class="constraints-${questionWithResponses.question.qnIndx}"></div>
```
on second thoughts I was thinking maybe its better to have an explicit closing tag ? HTML5 recommends separate closing tags for non-void elements. I checked on the W3C validator and it complains on giving a self closing tag to div.

mg14777 commented:
@wenmogu Ready for review. I have added some more tests

mg14777 commented on src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html:
```diff
@@ -1,14 +1,6 @@
-<div class="col-sm-6 width-auto">
+<div id="constSumSubmissionFormOptionFragment-${questionIndex}" class="col-sm-6 width-auto">
```
I think this issue is due to the class settings of contrib question instead. This problem exists in the master branch as well. The reason it is apparent now is because I am displaying the text boxes on a new line. The problem occurs due to unnecessary padding being added to these text boxes. I think the intention was to remove that by adding `mobile-no-padding` as shown below but it is ineffective:

![image](https://user-images.githubusercontent.com/11634699/40841537-b6083448-65a2-11e8-9ac6-a048311cfbd8.png)

I can fix this rn but I was wondering if this should be part of another issue ?

mg14777 replied:
I can add another css rule to override the default padding of columns ?

wenmogu replied:
there are many ways to tackle this problem i suppose. What we want is just the indentation for the label and input box to be consistent across different question types.

mg14777 commented:
@wenmogu I have added comments and regarding the padding issue I have commented above

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -475,19 +475,77 @@ function prepareConstSumQuestions() {
+    const constSumDistributePointsOptionSelected = $(`#constSumDistributePointsOptions-${qnNum}`).val();
+    const forceAllUnevenDistribution = constSumDistributePointsOptionSelected
+            === ParamsNames.FEEDBACK_QUESTION_CONSTSUMALLUNEVENDISTRIBUTION;
+    const forceSomeUnevenDistribution = constSumDistributePointsOptionSelected
```
`Specifies if at least one allocation must differ from the rest` is good

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -525,20 +559,27 @@ function updateConstSumQnInstructions(qnNum) {
         points *= numOptions;
     }
 
+    // Specifies the type of distribution constraint selected by question creator
```
`Specifies if the constraint for uneven distribution of points is selected` or something simpler

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -525,20 +559,27 @@ function updateConstSumQnInstructions(qnNum) {
 
     instruction += `${infoIcon} Total points distributed should add up to ${points}.<br>`;
+
+    // Check if allocating different points to just one recipient/option will be enough
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -525,20 +559,27 @@ function updateConstSumQnInstructions(qnNum) {
         instruction += `${infoIcon} At least one ${distributeToRecipients ? 'recipient' : 'option'} `
                 + 'should be allocated different number of points.<br>';
     } else if (forceUnevenDistribution || forceAllUnevenDistribution) {
+        // Each recipient/option must be allocated different points otherwise
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -573,7 +616,7 @@ function updateConstSumQnMessages(qnNum) {
     let allUnique = true;
     let allSame = true;
-    let allNotNumbers = true;
+    let allInputsEmpty = true;
```
nice naming

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -591,8 +634,12 @@ function updateConstSumQnMessages(qnNum) {
         let allChecksPassed = true;
-        if (!allNotNumbers) {
+
+        // Check if at least one input box is filled before displaying messages
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -621,7 +668,9 @@ function updateConstSumQnMessages(qnNum) {
                 allChecksPassed = false;
             }
 
+            // Check if allocating different points to just one recipient/option will be enough
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -621,7 +668,9 @@ function updateConstSumQnMessages(qnNum) {
 
+            // Check if allocating different points to just one recipient/option will be enough
             if (forceSomeUnevenDistribution) {
+                // display error if all allocated points are same when recipients are more than one
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,35 +611,49 @@ function updateConstSumMessageQn(qnNum) {
+                            + `${distributeToRecipients ? 'recipient' : 'option'} `
+                            + 'has been allocated different number of points.</span><br>';
+                }
+            } else if (forceUnevenDistribution || forceAllUnevenDistribution) {
```
comments here? since this is the part where the backward integration is needed, i think its better if we can hv some comments

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -645,10 +695,13 @@ function updateConstSumQnMessages(qnNum) {
                 }
             }
         }
+        // Check if all constraints have been satisfied
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -645,10 +695,13 @@ function updateConstSumQnMessages(qnNum) {
         }
+        // Check if all constraints have been satisfied
         if (allChecksPassed) {
+            // Change html class to green so that responses can be submitted successfully
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -645,10 +695,13 @@ function updateConstSumQnMessages(qnNum) {
             messageElement.addClass('text-color-green');
             messageElement.removeClass('text-color-red');
         } else {
+            // Change html class to red to prevent user from submitting an invalid response
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -591,8 +634,12 @@ function updateConstSumQnMessages(qnNum) {
         const approvedIcon = '<span class="glyphicon glyphicon-ok padding-right-10px"></span>';
         const errorIcon = '<span class="glyphicon glyphicon-remove padding-right-10px"></span>';
+
+        // Tracks whether all instructions/constraints have been satisfied
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -473,31 +473,130 @@ function prepareConstSumQuestions() {
+    }
+
+    // Specifies the type of distribution constraint selected by question creator
+    const constSumDistributePointsOptionSelected = $(`#constSumDistributePointsOptions-${qnNum}`).val();
```
Specifies if the constraint for uneven distribution of points is selected or something simpler

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -473,31 +473,130 @@ function prepareConstSumQuestions() {
+    // Specifies whether all recipients/options should be allocated different points
+    const forceAllUnevenDistribution = constSumDistributePointsOptionSelected
+            === ParamsNames.FEEDBACK_QUESTION_CONSTSUMALLUNEVENDISTRIBUTION;
+    // Specifies if all points can't be the same i.e. at least one allocation must differ from the rest
```
`Specifies at least one allocation must differ from the rest` is good

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -473,31 +473,130 @@ function prepareConstSumQuestions() {
+
+    instruction += `${infoIcon} Total points distributed should add up to ${points}.<br>`;
+
+    // Check if allocating different points to just one recipient/option will be enough
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -473,31 +473,130 @@ function prepareConstSumQuestions() {
+        instruction += `${infoIcon} At least one ${distributeToRecipients ? 'recipient' : 'option'} `
+                + 'should be allocated different number of points.<br>';
+    } else if (forceUnevenDistribution || forceAllUnevenDistribution) {
+        // Each recipient/option must be allocated different points otherwise
```
i think this comment is unnecessary

wenmogu replied:
and this part is where the backward integration variable comes in. You shld explain it.

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -517,34 +616,32 @@ function updateConstSumMessageQn(qnNum) {
+    /*
+     * Check current input provided by user
+     * and update error/sucess messages based
+     * on constraints
```
u can merge this line with the previous line

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -517,34 +616,32 @@ function updateConstSumMessageQn(qnNum) {
-                messageElement.removeClass('text-color-red');
-                messageElement.removeClass('text-color-blue');
+        // Tracks whether all instructions/constraints have been satisfied
+        let allChecksPassed = true;
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,39 +658,59 @@ function updateConstSumMessageQn(qnNum) {
-            message += ' Different amount of points should be given to some options.';
-            messageElement.addClass('text-color-red');
-            messageElement.removeClass('text-color-green');
+            // Check if allocating different points to just one recipient/option will be enough
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,39 +658,59 @@ function updateConstSumMessageQn(qnNum) {
-            messageElement.removeClass('text-color-green');
+            // Check if allocating different points to just one recipient/option will be enough
+            if (forceSomeUnevenDistribution) {
+                // display error if all allocated points are same when recipients are more than one
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,39 +658,59 @@ function updateConstSumMessageQn(qnNum) {
+                            + 'has been allocated different number of points.</span><br>';
+                }
+            } else if (forceUnevenDistribution || forceAllUnevenDistribution) {
+                // Display success message if all points are different
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,39 +658,59 @@ function updateConstSumMessageQn(qnNum) {
-        if (!allNotNumbers && ((forceUnevenDistribution && !forceSomeUnevenDistribution) || forceAllUnevenDistribution)
-                && !allUnique) {
-            message += ' The same amount of points should not be given multiple times.';
+        // Check if all constraints have been satisfied
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,39 +658,59 @@ function updateConstSumMessageQn(qnNum) {
-            message += ' The same amount of points should not be given multiple times.';
+        // Check if all constraints have been satisfied
+        if (allChecksPassed) {
+            // Change html class to green so that responses can be submitted successfully
```
i think this comment is unnecessary

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -561,39 +658,59 @@ function updateConstSumMessageQn(qnNum) {
+            messageElement.addClass('text-color-green');
+            messageElement.removeClass('text-color-red');
+        } else {
+            // Change html class to red to prevent user from submitting an invalid response
```
i think this comment is unnecessary

mg14777 commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -140,6 +140,10 @@ Apply to alert to override Bootstrap's default 3d alerts
 
+div[class*="col-"][class*="col-no-padding"] {
+    padding: 0;
+}
```
I added a new generic CSS rule which can be used anywhere when we want to remove the padding that comes along with a bootstrap column

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -574,12 +568,11 @@ function updateConstSumQnInstructions(qnNum) {
                 + 'should be allocated different number of points.<br>';
     } else if (forceUnevenDistribution || forceAllUnevenDistribution) {
-        // Each recipient/option must be allocated different points otherwise
+    // All points need to be different. Use forceUnevenDistribution for older versions to check the same
```
umm so this is how i have explained it atm. Let me know if something more needs to go in here.

mg14777 commented:
@wenmogu Ready for review.

damithc commented:
@mg14777 Instructions shown above the entry boxes are blue or grey?

mg14777 commented:
@damithc They are in blue

damithc commented:
> @damithc They are in blue

OK. I asked because #8864 shows the instructions in grey.

mg14777 commented:
@damithc Yes I had a feeling that's why you might be asking. The reason why that is happening is because Rank questions are passing those instructions from the server side by implementing the method `getInstructions()` and they get displayed by the existing instructions block in the HTML template.

For Const Sum questions the instructions returned were null originally for some reason which is why I added a separate `div` in the template and was manipulating instructions purely on the client side. 

I was wondering if it might be better to permanently change the formatting of the original instructions block (color them blue, with bullets etc.) and pass all these from the server side eliminating any client side manipulation ? That would universally affect all the question types though.

damithc commented:
> The reason why that is happening is because Rank questions are passing those instructions from the server side by implementing the method getInstructions() and they get displayed by the existing instructions block in the HTML template.

Do we know why the two question types behave differently? Should they be different?

mg14777 commented:
Ok I think I see what's going on. I was looking through the history of the method `getInstructions()` and it was originally introduced in PR [8037](https://github.com/TEAMMATES/teammates/pull/8037) for implementation of min/max feature for rank questions. The method was added in the parent class `FeedbackQuestionDetails` and only implemented for rank questions child class (files changed [here](https://github.com/TEAMMATES/teammates/pull/8037/files)) . For all other questions it was simply made to return null at the time and this never changed.

wenmogu commented on src/main/resources/feedbackQuestionContribSubmissionFormTemplate.html:
```diff
@@ -1,9 +1,9 @@
   </select>
 </div>
-<div>
+<div class="col-sm-6">
```
since u changed this part, probably (if u want to i.e. if u think this indent is a bit ugly, if not its fine) you will want to remove this indent of the more info link. 

![image](https://user-images.githubusercontent.com/28526038/40930271-c5177070-6859-11e8-8fa1-0da8df24dba2.png)

mg14777 replied:
forgot to add mobile no padding class. Will change that

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -473,31 +473,123 @@ function prepareConstSumQuestions() {
+        instruction += `${infoIcon} At least one ${distributeToRecipients ? 'recipient' : 'option'} `
+                + 'should be allocated different number of points.<br>';
+    } else if (forceUnevenDistribution || forceAllUnevenDistribution) {
+    // All points need to be different. Use forceUnevenDistribution for older versions to check the same
```
I think we can go with this ?
`// All points need to be different. forceUnevenDistribution is used to check the same for older versions
which did not contain the feature "some receive different points". In such versions forceUnevenDistribution
will be true for questions where "all receive different points" is selected.`

mg14777 commented:
@LiHaoTan @wenmogu Can this be merged now ? I can't finish the other referenced PR until this one is merged :sweat_smile:

damithc commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -517,34 +613,30 @@ function updateConstSumMessageQn(qnNum) {
     }
-
+    /*
+     * Check current input provided by user
```
Checks ...

mg14777 commented:
@LiHaoTan @wenmogu Fixed indentation and typos
